Riscv port

ct-clmsn opened the pull request:
### Ticket

Link to Github Issue #14909 

### Problem description

Provides preliminary support for rv64g sifive unmatched hardware.

This patch generalizes several existing x86 specific portions of the tt-metal stack such that they are now portable across multiple architectures - the x86 intrinsics have been ported to ISO C++.

Part of this port modifies the existing build script so that users can now select different C and C++ compilers (and versions of compilers) to build tt-metal.  The selected compilers properly forward into to the underlying tt-metal CMake infrastructure.

Thanks to @joelsmithTT for making time for discussion and verification.

### What's changed

Small portions of the existing C++ functionality use x86 specific intrinsics. This patch modifies those portions in order to generalize them to ISO C++. This work relied heavily on the following [link](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html) which describes in general terms (pseudo code) what operations a particular set of x86 intrinsics perform.

The bash script that runs the build has been updated to allow users to select compilers (C/C++).

A substantial number of unit tests pass but not all of them. There are several instances where a unit test composed of several other unit tests pass but, when the dependency unit tests are executed  they fail with a segmentation fault.

Some tests also seem to require `TT_METAL_SLOW_DISPATCH_MODE=1` but do not state it's a requirement unless they are run independently. As an example, running the "unit_test" program will run several unit tests without that environment variable activated. Several of the tests run as part of "unit_test" will fail. Some of tests that fail can run correctly with that environment variable outside of being run as part of the umbrella suite found in "unit_test".

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

mo-tenstorrent committed:
#9956: Support perf report for tracing runs

o2buzzle committed:
#13523: added expand op (for row-major tensors)

punithsekar committed:
#13053: Yolov4 cookbook

tt-dma committed:
#0: DPRINT TileSlice refactor/cleanup

Also implement Float32 printing

Also adds first-cut implementation of dprinting float32

tt-dma committed:
#0: Update DPrint error messages

o2buzzle committed:
#13523: tile "support" for expand

abhullar-tt committed:
#0: remove unused defines maps from JitBuildOptions

tt-dma committed:
#13179: Add support for DPRINTing Bfp8_b and Bfp4_b tiles

Also add a directed test for TileSlice printing

pgkeller committed:
#11795: Update test_pgm_dispatch and sweep

Add dispatch to eth tests
Add dispatch from eth flag
Add long running + shadow kernel tests
Add a combo of the above

TT-BrianLiu committed:
#13127: Add compute_output_specs which allows user to return output tensor layout
- Op must implement either compute_output_shapes + create_output_tensors OR compute_output_specs
  * If compute_output_specs is implemented, infra will handle tensor creation
  * Also works for optional output tensors
- Update generic reduction which carries over padding along N and C
  * pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_min_max.py::test_min_max_for_dim_hw
  * pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_sharded.py::test_sharded_reduce_h
- Update transpose op which transposes and maintains padding
  * pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_transpose.py
- Update slice op which uses optional output tensors
  * pytest tests/ttnn/unit_tests/operations/test_slice.py::test_slice_output_tensor_tile

vigneshkeerthivasanx committed:
#13326: move ttnn reshape to device on vgg model

mo-tenstorrent committed:
#6802: Enable buffer usage monitoring in tracy

mo-tenstorrent committed:
#0: Comment out Zonescopes on program build flow

Commit:
#14083: Remove compiler warning caused by PT_RISCV_ATTRIBUTES

abhullar-tt committed:
#10973: Load and launch FW on both BH eth core riscs

o2buzzle committed:
#13523: remove unused kernels

Commit:
#13644: Add support for tensor-scalar binary ops

abhullar-tt committed:
#10973: Add data movement processor to ethernet kernel config

caixunshiren committed:
#14103: improved multi-core parallelization in flash decode

esmalTT committed:
#14283: Add tests for UNet Shallow grouped convolutions

mywoodstock committed:
#14260: maxpool bugfix -- when data is odd, need to pad cores

umadevimcw committed:
#13758: Update logical xor op and remove composite version

shwetankTT committed:
#10110: adding comment for use_non_height tile.

umadevimcw committed:
#13758: Update logical op documentation

esmalTT committed:
#0: Assert on sharded+tiled HC transpose inputs

Sharded and tilized inputs are not currently supported in HC transpose.
Previously, this condition was not checked and the operation produced
garbage outputs.

TT-billteng committed:
Revert "#0: SFPU LLK Unit Test Parameter Sweep" (#14402)

Revert "#0: SFPU LLK Unit Test Parameter Sweep (#14218)"

This reverts commit ad61399224ea969909886b4fd8964bb8ecafd15e.

tt-asaigal committed:
#0: Move SynchronizeWorkerThreads to tt_metal::detail namespace (#14322)

- This is a core tt_metal API and shouldn't be in tensor_ops
- Also account for worker thread deadlock when calling this API

Commit:
#14361: Implement program API

blozano-tt committed:
#0: CMake: Remove stdlib interface library (#14320)

abhullar-tt committed:
#10973: Adding BH tests for running kernels on both eth riscs

Aswinmcw committed:
#14267: Move sender receiver computation to ccl_common.cpp (#14269)

* #14267: Move sender receiver computation to ccl_common.cpp

* #14267: Use snake_case

sminakov-tt committed:
#14363: Optimize Shape::operator[] (#14399)

* #14363: Optimize Shape::operator[]

* #0: Negative index fix

Aswinmcw committed:
#5560: `all_reduce` support (reduce_scatter + all_gather) (#14085)

* #5560: Initial commit to get reduce_scatter as common

* #5560: Modify test file

* #5560: Add optimized version

* #5560: Add enum for strategy

* #5560: Use static methods

* #5560: Update namespace

* #5560: Update pytest constants

* #0: Rebased

Commit:
#13774: Fix subalpha_bw op (#14341)

ngohoang34 committed:
#14134: Add bfloat8_b unit test for moreh_layer_norm (forward, backward) (#14207)

#14087: Add bfloat8_b unit test for moreh_layer_norm (forward, backward)

nemanjagrujic committed:
#11512: sub_bw, fill_zero_bw, log_sigmoid_bw (#13805)

* #11512: Add sweeps for sub_bw, fill_zero_bw, log_sigmoid_bw
* #11512: Add sweeps for ldexp_bw, logaddexp_bw, hardshrink_bw, logit_bw, neg_bw, softhhrink_bw
* #11512: Add sweep for addcmul_bw
* #11512: Add sweeps for complex conj, complex reciprocal and complex reciprocal_bw

Commit:
#13794: Conv2d BS arbitrary kernel dims

Allow arbitrary kernel dimensions in block
sharded conv2d.

This enabled more torch traces to pass as
block sharding is now more viable option is
auto-shard codepath for convs.

Logic for arbitrary double buffer of activations
and weights in block sharded code path is removed.
This caused issues with some torch trace examples
that used to pass with height sharding, as now
auto shard would pick block sharding as a better
option, but arbitrary double buffer would cause
out-of-memory issues.

KalaivaniMCW committed:
#14407: Add channel bcast support for non-4D tensors (#14408)

nathan-TT committed:
#12979: Explict phdrs, enforce ordering (#14359)

nathan-TT committed:
#12979: Remove unused l1_memory section (#14304)

TT-billteng committed:
Update timeout limit of C++ tests to 60 minutes (#14429)

tt-aho committed:
#0: Add new contains/intersects apis for cores
Change CoreRangeSet constructor to take in a tt::stl::Span
Rename core_coord_in_core_ranges to contains to match other apis

tt-aho committed:
#13655: Add limited ability to shrink allocator size from the bottom

yugi957 committed:
add sweeps for embedding, repeat, split, squeeze, stack, unsqueeze, view (#14385)

* #13181: added sweep for squeeze_pytorch2

* #13181: sweep only for model traces

* #11621: merged yaml changes

* #11621: merged embedding sweep changes

* #11621: merging view changes

* #11621: merging sweep changes

* #11621: add squeeze workflow

* #11621: view_trace md

* #11621: embedding pcc accuracy assertion increased, sweep has uint32 for indices now

mo-tenstorrent committed:
#9956: Trace profiling smoke test

blozano-tt committed:
#14329: Add MEM_MAILBOX_BASE to Hal (#14362)

pgardnerTT committed:
#5424: Cleanup lrelu (#13959)

* #5424: Cleanup lrelu

SeanNijjar committed:
#0: update ccl test owners (#14357)

blozano-tt committed:
#12743: Fix namespace tt::tt_metal usage in tt_metal (#14413)

cglagovichTT committed:
Llama3.2-Vision: Add functionality for Llama-Vision demo and tests (#14051)

* #13368: Move repeat interleave to xattn cache generation.

* #0: Clean up demo, enable arbitrary padding for multimodal text sequence

* #13368: Add llama_models Meta reference for Llama3.2 as a submodule

* #13368: Change reference imports to use new submodule

* #13368: Clean up comments after pushing repeat_interleave into xattn_cache generation.

* #13368: Clean up vision tests. Unify assertions and pcc checks. Fix LM head splitting on T3k.

* #13368: Fix LM head splits calculation

* #13368: For all vision tests, get model-specific parameters from model_args rather than fixtures. This generalizes tests for base and finetuned 11B models.

* #13368: Add vision tests to unit, frequent, and demo

* #13368: Fixup mesh_device when not passed FAKE_DEVICE

* #13368: Remove llama_models as submodule. Move its install to llama3 requirements.txt.

---------

Co-authored-by: mtairum <mtairum@tenstorrent.com>

tt-rkim committed:
#14440: Skip mistral7b on single-card WH demos as it seems to kill WHs for now [skip ci]

blozano-tt committed:
#14367: Inline operator[] Again for RuntimeArgsData (#14384)

TT-billteng committed:
Enable weka on models post commit workflow (#14431)

#0: enable weka on models post commit workflow

blozano-tt committed:
#12743: Enable clang-tidy bot to run one check (#14438)

sminakov-tt committed:
#14366: Allow creating pre-allocated buffer with an address. Minor perf improvements (#14394)

* #14366: Allow creating pre-allocated buffer with an address. Minor performance improvements

* #0: Cleanup

* #0: Updated methods documentation

* #0: Handle allocation failure

* #0: Scheduling fix

hschoi4448 committed:
#14453: Fix `moreh_getitem` assert bug (#14454)

#14453: fix moreh_getitem assert bug

Commit:
#13864: Add forward support for maximum  and minimum with tensor, scalar variant (#14266)

* #13864: Add forward support for maximum op with tensor, scalar variant

* #13864: Add forward support for minimum op with tensor, scalar variant

nilaykp committed:
Make reductin on 32 rows at a time instead of 16. (#14344)

#12507  Make reductin on 32 rows at a time instead of 16.

Signed-off-by: Nilaykumar K Patel <nkpatel@tenstorrent.com>

BuiChiTrung committed:
#14262: add bfp8 test for moreh_nll_loss and moreh_adamw (#14263)

* #14262: add bfp8 test for moreh_nll_loss and moreh_adamw

* #14262: use from_torch in adamw

* #14262: accept bfp8 in moreh adamw

* #14262: rollback umd

* #14262: update test

thanhnguyen1512 committed:
#14132: Add BFP8 unit tests for moreh dot (#14143)

* #14132: Add BFP8 unit tests for moreh dot

* #14132: Add BFP8 unit tests for moreh softmax

* #14132: Add BFP8 unit tests for moreh softmax backward

* #14312: Removed print() and used to_torch() for output

* #14132: Reformat tests

yieldthought committed:
#3323: Disable mistral7b demo and add warning to README

Commit:
#14462: Conv2d account for split reader in determine_per_core_conv_block_config

Commit:
#12815: update mish op (#14270)

Commit:
#14464: Re-enable split_reader on gs resnet

After commit #0194005 Resnet50 device perf
regressed to 6600.
Commit above decoupled two conv2d features
shallow_conv and split_reader.
Previously if shallow conv was enabled,
split_reader would be enabled under the hood even
if it's not requested.
Resnet50 on wh was configured to use split_reader,
but on gs it was configured not to use split_reader
and this lead to perf regression.

This change configures split_reader on gs explicitly
to restore the perf.

TT-BrianLiu committed:
#13127: Add helper function and tests for converting between 2D logical data and 2D physical data
- Process for converting logical to physical is:
  1. Logical shard shape slices 2D logical data
  2. Each shard is padded up to physical shard size
- Process for converting logical to physical is:
  1. Strip padding for each shard based on where logical shards are
  2. Combine logical shards into 2D logical data
- Build with ttnn unit tests and run:
  * ./build_Release/test/ttnn/unit_tests_ttnn_tensor --gtest_filter=TensorShardingTests/ShardWithAlignmentTests.LogicalToPhysical*
  * ./build_Release/test/ttnn/unit_tests_ttnn_tensor --gtest_filter=TensorShardingTests/ShardWithAlignmentTests.PhysicalToLogical*

tt-aho committed:
#14457: [skip ci] Fix CreateSemaphore documentation related to return value (#14473)

#14457: Fix CreateSemaphore documentation related to return value, as well as number and size of semaphores

mcw-anasuya committed:
#13626: Update doc for lgamma, normalize_global, normalize_hw, mish (#14214)

* #13626: Update doc and sweep test for lgamma

* #13626: Update normalize_hw, normalize_global

* #13626: Update doc for mish

* #13626: Update range for rpow

Aswinmcw committed:
#14406: CCL Perf Monitoring (#14420)

* #14406: Add initial script to get perf report

* #14406: Add reference values and assert

* #14406: Modify summary to show for multiple runs

* #0: Update copyright year

abhullar-tt committed:
#0: [skip_ci] Adding allocator tech report

mcw-anasuya committed:
#14448: Update examples of binary ops (#14468)

#14448: Update examples, wordings of binary ops

blozano-tt committed:
#0: Remove dead code in CMakeLists.txt (#14478)

afuller-TT committed:
#0: Bump the timeout (#14482)

Successful runs already brush up against this limit: https://github.com/tenstorrent/tt-metal/actions/runs/11587903090/job/32260805419
And some runs trip over it: https://github.com/tenstorrent/tt-metal/actions/runs/11595963824/job/32286007664

abhullar-tt committed:
#14456: Fix for pad on BH

blozano-tt committed:
#0: Remove references to ARCH::WORMHOLE (#14469)

tt-rkim committed:
#12626: [skip ci] Add more detailed comment of why we need no --log-driver (#14488)

Force-merging for documentation

tt-rkim committed:
#14484: [skip ci] Get rid of spurious comments on PRs about "deploying" by erasing `environment: ` usage in workflows (#14490)

* #14484: Remove dev environment as it contains environment variables that are not used anywhere and it clutters PRs

* #14484: Remove unnecessary comment

jvegaTT committed:
#12516: Implementing a Barrier CCL OP (#14036)

* #12516 Created non-device connection for barrier implementation

* #12516: Made ping latency uBenchmark into OP

* #0: Bug fix on hop latencies uBenchmark

Runtime arguments were flipped between semaphore and Buffer

* #0: Update Programming Mesh of Devices with TT-NN.md

* #12516: Added barrier op, has ethernet semaphore issue

Adding the barrier operation but it fails due to the fact ethernet semaphore does not initialize correctly

* #12516: First working barrier
Need to gauge performance and effectiveness

* #12516 Created non-device connection for barrier implementation

* #12516: Made ping latency uBenchmark into OP

* #0: Bug fix on hop latencies uBenchmark

Runtime arguments were flipped between semaphore and Buffer

* #12516: Added barrier op, has ethernet semaphore issue

Adding the barrier operation but it fails due to the fact ethernet semaphore does not initialize correctly

* #0: Update Programming Mesh of Devices with TT-NN.md

* #12516: First working barrier
Need to gauge performance and effectiveness

* #12516: Added nightly and post commit testing to the barrier op

* #12516: Moving barrier testing from nightly to t3000 frequent

* #0: Correction to Bug fix on hop latencies uBenchmark commit

Also rotating the semaphore in the sender, not just in the receiver

* #0: Adding license information to barrier_sem_creator.cpp

* #0: license information added to barrier_full_worker_grid.cpp

* #0: Fixing licensing bug

* #0: Making buffer region a multiple of 32

* #12516: Modifying dataflow_api.h for performance

Creating the send_eth_receiver_channnel_done for when the eth_channel_sync_t object is already known
Will spin loop a bit before calling run_routing() for semaphore waits to prevent perf loss from over-calling run_routing()
Formatting changes for comments

* #12516: Applying PR recommendations

* #12516: Removing now depricated test_barrier.py

* #12516: Adding test validation for all gather in sharded test

* #12516: Moving from launch_op to invoke

* #12516: Other suggested changes

* #12516: Reusing pre-made handshake API

* #12516: Moving as much as possible pre-processing from launch_op to invoke

* #12516: Removed redundant send_eth_receiver_channel_done which is now part of the api

* #12516: Added pyskip for n300

* #0: Removing blank space in C make lists

* #12516: Improvements to barrier op and dataflow_api.h

Created an eth_wait_for_bytes_on_channel_sync_addr function for when the channel buffer is not one of the defaults in erisc_info->channels
Added wait_for_eth_txq_cmd_space to properly wait for the txq function to finish rather than using an empty while loop
Moved the barrier frequent test to the ccl folder
Improvements on the barrier op using new aforementioned functions

* #12516: Updated run_t3000_frequent_tests.sh with new location of ccl tests

* #12516: Move rt_args from std::vector to std::array

* #12516: changing get_t3k_physical_devices_ids_ring to open_mesh_device as recommended

* #12516: Applying api recommendations

Using ttnn.synchronize_devices instead of synchronize_device

Using from_torch to create the tensor mesh in one step

* #12516: Minor documentation error on barrier OP

* #12516: removing unused libraries in barrier.hpp

* #12516: Removing barrier_op.hpp inclusion in barrier.hpp

* #12516: Changing function name from barrier to barrier_function

* #12516: Correcting clang-diagnostics-error

* #12516: Other clang errors fix

Commit:
#14494: Implement tt::stl::AnyRange

tt-aho committed:
#0: Update num_cores_to_corerange_set to return a CoreRangeSet and rename to num_cores_to_corerangeset
Split work_split.hpp into impl and header file

kpaigwar committed:
#0: integrated sharded reduce scatter in TG Model (#14467)

#0: integrated sharded reduce scatter

ayerofieiev-tt committed:
Introduce TensorLayout in Tensor implementation (#13607)

DuongQLee committed:
#13268: Implement fold (#13924)

#13268: initial commit

o2buzzle committed:
#13523: fix compilation issues

dmakoviichuk-tt committed:
Fixed header file for unity build (#14504)

#0: fixed header file for unity build

TT-billteng committed:
Bump timeout limit of SD unit tests to 35 minutes (#14506)

o2buzzle committed:
#14508: Add bfloat8_b tests for moreh_adam, moreh_matmul[_backward] and moreh_sum[_backward]

amalbasaTT committed:
Refactoring unary_backward ops sweeps (#13942)

* #11512: refactoring unary backward sweeps

* #11512: Modify pow_bw sweep parameter

* #11512: Rebasing refactoring branch

* #11512: Rebase refactoring branch

* #11512: Rebasin refactoring branch

* #11512: Rebasin refactoring branch

* #11512: Remove submodule update from refactoring branch

* #11512: Move softplus_bw to xfail

DuongQLee committed:
#14155: Add bfloat8_b tests for moreh_sgd, moreh_linear, moreh_linear_backward (#14156)

* #14155: add bfp8_b tests for moreh_sgd, moreh_linear and moreh_linear_backward

* #14155: remove create_ttnn_tilized_tensor

amalbasaTT committed:
Add transformer sweeps (#14426)

* #11512: Add concat_heads and split_qkv_heads sweeps

* #11512: add other version of split_qkv_heads sweep and modify ttnn-run-sweeps yaml

* #11512: Add split_qkv_head third sweep

* #11512: Refactor split_qkv_heads sweeps

* #11512: Refactor split_qkv_heads sweeps

* #11512: Refactor split_qkv_heads sweeps

* #11512: Refactor split_qkv_heads sweeps

* #11512: trasformers sweeps minor fix

jdh8 committed:
#13337: Optimize trigonometric functions (#14368)

* #13337: Optimize sine and cosine

The coefficients in this commit were generated by:
https://gitlab.inria.fr/sfilip/rminimax

That tool is used in the CORE-MATH project, a library of correctly rounded mathematical functions.
https://core-math.gitlabpages.inria.fr/

* #13337: Extend domain of `ttnn.tan`

* #13337: Improve accuracy of tangent

* #13337: Update to deterministic rounding mode to `float_to_int16` in trigonometric kernels

* #13337: Also apply trig optimization to Blackhole

* #13337: Improve precision of tan over [-1, 1]

`sfpu_reciprocal` is still inaccurate.  We need another issue for this.

* #13337: Calculate `ttnn.tan` without using `reciprocal` for better precision

---------

Co-authored-by: Artem Yerofieiev <169092593+ayerofieiev-tt@users.noreply.github.com>

yugaoTT committed:
Add matmul_2d benchmark test  (#14483)

* #0: add matmul 2d benchmark test

* #0: add bfloat8 matmul benchmark test

* #0: only run for whb0

* #0: skip test due to didt

sagarwalTT committed:
#11795: Implemented randomized FD tests with trace to test the ring buffer (#14485)

* #0: Initial commit

* #0: Added trace tests

* #0: Run trace for several iterations

sjameelTT committed:
#14372: fix transpose hc RM pcc errors + add N-dimensional permute single core implementation (#14388)

#14372: fix transpose hc RM pcc errors
- Restriction was on alignment, which was hard coded to 16 for L1, which caused PCC issues when read from DRAM
#14370: add basic N-d permute code and support both N-d transpose and permute
- TODO: make multicore, single-core for now

ncvetkovicTT committed:
#14510: Disable reduce scalar until investigated (#14523)

ntarafdar committed:
#14324 Add Uint32 Support for Untilize and Tilize

* #0: add uint32 support to untilize and tilize

afuller-TT committed:
#0: Workflow to test compatibility with TTNN Project Template (#14442)

blozano-tt committed:
#0: Ensure libc++ used across both projects (#14254)

Problem description
Fix linker problem
UMD does not ingest stdlib interface library from tt-metal anymore.
Need to get the two projects to use same version of C++ standard library.
Solution is to adjust CXX_FLAGS.

What's changed
Remove stdlib interface library, use CXX_FLAGS.
Also removed DFMT_HEADER_ONLY define since it comes for free, when you link the right target.
Had to fix a couple tangential issues that popped up.
---------

Co-authored-by: Bojan Rosko <brosko@tenstorrent.com>
Co-authored-by: Bojan Roško <156314064+broskoTT@users.noreply.github.com>

afuller-TT committed:
#14001: Tidy up tt_metal's CMake (#14371)

* No need to extract object files from Object libs (not since CMake v3.14)
* Create a namespaced ALIAS and update consumers to use it
  * Using a namespaced target gives more info to CMake to detect mistakes earlier and with a better diagnostic message
* Remove noise in `add_subdirectory` calls
  * Reduces cognitive load when reading; no need to scrutinize for typos, and deviations from the norm stand out like a sore thumb rather than hide in the haystack.
* Collapse both `target_link_libraries` calls into one
* Remove an unnecessary level of indirection (a variable that is used only once and adds no additional clarity)
* Move the `add_subdirectory` calls to the end
  * The focus should be on THIS level, and lower levels are implementation details

yugi957 committed:
Added absolute path for view trace for CI (#14534)

#11621: test absolute path for correct CI for view

mbezuljTT committed:
#0: fix broken tt-mlir build (#14522)

* #0: fix broken tt-mlir build, fsized-deallocation is required for a path when compiler is clang but libcxx is disabled

afuller-TT committed:
#0: Use Unity build (#14535)

TT-billteng committed:
Update build command used by bisect script (#14544)

#0: update build command used by bisect script

LPanosTT committed:
Expose conv2d weight/bias preparation as ops (#14049)

Added new weight and bias preparation ops, and new conv op that can only take pre-prepared weights.
Working conv test with pre-prepared weights

added return weight/output dims kwargs

Only auto-shard if shard_layout not specified

Pass input memory config to prepare functions

Organize utility functions into their own files

abhullar-tt committed:
#14518: Skip unsupported repeat case for BH

avoraTT committed:
#14107: Add support for decode shapes in RoPE op (#14241)

* #0: Changed RoPE pytest to test for decode shape. Added the ttnn.embedding to select specific sin/cos. TODO: Change RoPE kernel code to suppport decode shapes that are sharded.

* #0: Add new flag to rope op to differentiate between prefill/decode. Use get_logical_shape in validation code.

ncvetkovicTT committed:
#0: Workaround for #14510, disable all reduce scalar (#14555)

nemanjagrujic committed:
Ngrujic/sweep tests 2 (#14290)

* #11512: Add sweep test for ttnn.prod
* #11512: Add sweep test for ttnn.sum
* #11512: Add sweep test for ttnn.l1_loss and ttnn.mse_loss
* #11512: Add sweep test for ttnn.reshape
* #11512: Add sweep test for ttnn.repeat_interleave

hschoi4448 committed:
#14316: Refactoring moreh_helper function (#14317)

* #14316: refactoring moreh_helper

arikTT committed:
#13637: Enable L-shape multicast API and add tests (#14309)

* #13637: Adding L-shape multicast - low level API, test. Still WIP

* #13637: Added L-shape multicast API and tests

* #13637: update docs and not limit tests to slow dispatch

* #13637: fix pre-commit failures

* #13637: clean up debug code

* #13637: fix noc 1 core calculations, wrap BH only code with #ifdef

* #13637: internally handle noc1 exclusion direction, update API method name

* #0: fix typo

npetrovic-tenstorrent committed:
#11512: Sweeps: addcdiv_bw, tilize, untilize (#14511)

* #11512: Add addcdiv bw sweep

* #11512: Add tilize sweeps

* #11512: Add untilize sweeps

* #11512: Fix run

* #11512: Fix run conflict

LPanosTT committed:
Revert "Expose conv2d weight/bias preparation as ops" (#14563)

Revert "Expose conv2d weight/bias preparation as ops (#14049)"

This reverts commit 09fd48cca7d31cef373b1b6488fa1e9f4cbdb986.

Force-merging to fix perf pipelines.

nhuang-tt committed:
#14382: fix invalid lengths in prefetcher test (#14481)

invalid lengths were being created by the test gen
functions

ensure that the lengths are aligned correctly and
bounded within the min/max sizes

jbaumanTT committed:
#14501: Stall enqueue_program after enqueue_trace (#14503)

* #14501: Stall enqueue_program after enqueue_trace

enqueue_program normally writes out the RTA to the ringbuffer before it
stalls waiting for the previous program to finish. If a kernel is
running due to enqueue_trace, the config buffer manager is unaware of the
memory that's in use by the kernel, so the RTA may overwrite the kernel
binary or other L1 memory that's in use.

The WorkerConfigBufferMgr should be informed that its state is out of
date, so it will cause a stall before writing out the next data.

ncvetkovicTT committed:
#0: Enable UnpackTilize Unit Test for BH (#14564)

* #0: Enable UnpackTilize* tests for BH

spoojaryTT committed:
Remove dependency of Kernel on Device. (#14561)

* #13869: Move noc_size and allocator decleration to core_config

* #13869: Remove usage of IS_NOT_POW2 from get_bank_index

* #13869: Fix blackhole allocator alignment value.

sminakov-tt committed:
#0: Fixes for TensorLayout fromLegacyPaddedShape (#14542)

* #0: Fixes for TensorLayout fromLegacyPaddedShape

* #0: Review fixes, cleanup

* #0: Non-zero page size

yugaoTT committed:
Change logger info for matmul benchmark (#14577)

* #0: change logger info

* #0: add comment for matmul benchmark

jbaumanTT committed:
Reland ring buffer (#14568)

* #111795: Revert "Revert Kernel Ring Buffer Changes (#14244)"

For tensix only (disabled for eth for now)
Enables execute in place (XIP) so kernels can be loaded anywhere
Moves kernels from FW address "holes" to (larger) ring buffer
Still runs synchronously.

This reverts commit e966f7768c9db0ed45cd8733831622b303f3e2c3.

abhullar-tt committed:
#14548: Update untilize with unpadding test params for BH

tt-aho committed:
#0: Add explicit codeowners to tt_metal/tt_stl

blozano-tt committed:
[skip ci] #0: Update INSTALLING.md with cmake version note (#14576)

ayerofieiev-tt committed:
#0: Handle 0 volume case in layout change code (#14578)

#0: handle 0 volume case in layout change code

afuller-TT committed:
#14001: De-globalize reflect (#14432)

### Ticket
#14001 

### Problem description
Fixing globals in service of being able to package up Metal.

### What's changed
Object-ify `reflect` and consume it where it's actually used, allowing it to propagate as needed.
Remove the global reference to `reflect`.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11632757225
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes
- [x] TTNN C++ Project passes https://github.com/tenstorrent/tt-metal/actions/runs/11618754311

abhullar-tt committed:
#14613: Remove eth cores from BH soc so they cannot be targeted while debug is in progress

blozano-tt committed:
#0: Consolidate all of the noc_parameters into metal (#14539)

VirdhatchaniKN committed:
#13967: Support for clip_bw (#14093)

* #13967: Support for clip_bw

* #13967: Update

* #13967: Update code

abhullar-tt committed:
#10973: Fixes for enabling second erisc after ring buffer changes

broskoTT committed:
Change tt_cluster to use single tt_SiliconDevice (#13949)

Ticket
First step towards #13948

What's changed
Changing tt_cluster so that it uses a single tt_SiliconDevice (to be renamed to tt::umd::Cluster) for all connected chips.
In UMD, the following changes were made prior to this change:

Some API tests added to track current API usage.
Made a test which explicitly shows our current driver doesn't support multiple separate clusters of chips (e.g. two N300 cards not connected through ethernet).
chip_id added to setup_core_to_tlb_map, since each mmio chip has it's own set of TLBs
Changed wait_for_non_mmio_flush to be able to accept chip_id, so it can run only for a specific remote chip.
chip_id added to get_pcie_base_addr_from_device
dynamic_tlb_config removed from constructor (this might be checked in with another PR before this one, since it is already in master)
Related UMD PRs
tenstorrent/tt-umd#166
tenstorrent/tt-umd#179
tenstorrent/tt-umd#165
tenstorrent/tt-umd#183
tenstorrent/tt-umd#131
tenstorrent/tt-umd#195
tenstorrent/tt-umd#247

abhullar-tt committed:
#0: Reduce idle erisc heartbeat posting that caused hang in running SD idle erisc kernels

blozano-tt committed:
#14631: Update device.hpp (#14632)

#0: Update device.hpp

abhullar-tt committed:
#0: Skip erisc0 test on WH until #14640 is addressed

shwetankTT committed:
#14401: yolo4 move from experimental to demo

johanna-rock-tt committed:
#0: Fix perf report: signposts with multi device (#14526)

Co-authored-by: Johanna Rock <jrock@tenstorrant.com>

Commit:
#14145: Update Documentation for backward ops (#14412)

Commit:
#12835: Fix binary fmod (#13568)

tt-rkim committed:
#0: Improve falcon7b device perf threshold because literally NO ONE IS FIXING IT

amalbasaTT committed:
sanitize_shape refactoring (#14565)

* #11512: sanitize_shape refactoring

* #11512: remove sanitize_topk_shape

* #11512: Remove print functions from sanitize_shapes

* #11512: Modify sanitize_shape_rm function

* #11512: Reduce argmax sweep number of parameters

* #11512: Modify invalidate_function inside split_query_key_valu_and_split_heads..py sweep

* #11512: Modify santizie_shape function inside utils when method is split_query_key_value_and_split_heads

abhullar-tt committed:
#0: Fix for running kernels out of idle erisc cores on WH

Commit:
#14494: Enhance AnyRange

sjameelTT committed:
#12549: Fix BH unaligned read issue for tiled interleaved transpose HC (#14572)

- Replace buffer.alignment() calls with buffer type check as
buffer.alignment() always returns DRAM alignment even though it's in L1
- Reenable some disabled test
- Fix resnet assert with the buffer.alignment() removal

shwetankTT committed:
#14401: yolo v4 on nighty CI.

derdeljanTT committed:
#12624: Fix tracking of globally allocated CBs in graph capture (#14521)

* #12624: Fix tracking of globally allocated CBs in graph capture

* #12624: Update graph tracing docs

skhorasganiTT committed:
[skip ci] Update perf and latest features for llm models (Nov 4) (#14680)

afuller-TT committed:
#14571: Check CMake compatibility (#14573)

Ticket
#14571

Problem description
We advertise a certain CMake compatibility, but our CI systems have a newer version for convenience. We need a quick check to ensure we don't slip up and put something in that requires a newer version.

What's changed
New check that uses a fixed (old) version of CMake and confirms whether it can generate build files.

blozano-tt committed:
#0: Update dprint_server.cpp (#14647)

blozano-tt committed:
#14638: Update tt_metal_profiler.cpp (#14639)

Update tt_metal_profiler.cpp

jbaumanTT committed:
#11795: Dispatch async to ringbuffer

Don't wait for previous programs to finish before copying new ones into the
ringbuffer. This allows more parts of dispatch to work in parallel with
programs.

Ethernet kernels aren't placed in the ringbuffer, so any programs including
ethernet kernels dispatch the same as previously.

Also move the WorkerConfigBufferMgr from the SystemMemoryManager to the
HWCommandQueue, since it relies on sync counts that are dispatcher-specific.
Between command-queues we need to wait for the commands to completely drain,
since we have no way to wait for a specific command in a different command
queue to finish.

abhullar-tt committed:
#12349: Enable working tests on BH

blozano-tt committed:
#14001: Bump UMD to help w/ CMake cleanup (#14687)

* #14001: Bump UMD to help w/ CMake cleanup

* #14001: Extra bump

---------

Co-authored-by: Andrew Fuller <afuller@tenstorrent.com>

VirdhatchaniKN committed:
#13234: Support clip, clamp with min/max tensor (#14660)

* #13234: Support clamping with min/max tensor

* #13234: Update

* #13234: Update

* #13234: Update clip

* #13234: Update clip

* #13234: Update pybind

KalaivaniMCW committed:
#14336: add bfloat8b support for _like creation ops (#14346)

Commit:
#13794: Relex auto-shard rules for block sharding

#d28568c5 added support for arbitrary kernel dims for block
sharded conv2d.
This commit relaxes rules for auto-shard codepath, so that
auto-shard heuristic can pick block-sharding for any
kernel dims.

johanna-rock-tt committed:
#14512: Add DRAM pre-fetcher benchmark (#14528)

* #14512: Add DRAM pre-fetcher benchmark

mcw-anasuya committed:
#11086: Update unary backward doc set 3 ops (#14670)

* #11086: Update doc for pow_bw

* #11086: bind_unary_backward_shape

* #11086: Update doc for abs_bw

Commit:
#12815: Fix nightly (#14663)

tt-aho committed:
#14741: Add implementation and basic tests for global semaphores

afuller-TT committed:
#0: Lint YAML files (#14696)

afuller-TT committed:
#0: Permit merges and --fixup (#14698)

pgkeller committed:
#0: Update bw and latency test/scripts

Sweep mcast
Make it easier to re-generate these numbers

mo-tenstorrent committed:
#9956: Device only perf report smoke test

mo-tenstorrent committed:
#9956: Better op report column names

blozano-tt committed:
#14636: Update device_command.hpp (#14709)

johanna-rock-tt committed:
Revert "#14512: Add DRAM pre-fetcher benchmark (#14528)" (#14749)

This reverts commit 6a9a66f5860f4b2eb31cc61f94eb0b50122e0db9.

ttmchiou committed:
#0: Adding additional codeowners t3k and galaxy workflows

Commit:
#12855: Fix unpacker race condition by putting stall after MMIO

blozano-tt committed:
[skip ci] #0: Update all-static-checks.yaml (#14759)

bbradelTT committed:
#14351: enable single core fused op tests on BH (#14761)

ttmchiou committed:
#14773: Set default to true when getting active ethernet cores

blozano-tt committed:
#14620: Remove unnecessary ARCH_NAME specific includes from llrt.hpp (#14621)

bbeggsTT committed:
Create Tutorial - Add Two Integers in a Baby RISC-V.md (#14527)

esmalTT committed:
#14830: Add support for groups in `ttnn.concat`

When `groups` is set to a value greater than 1, the inputs are split
into N `groups` partitions, and elements are interleaved from each group
into the output tensor. This feature is useful for recombining grouped
convolution outputs during residual concatenation.

The default behavior of `ttnn.concat` is unchanged.

umadevimcw committed:
#14712: Update unary documentation

* #14712: Update the unary bind operation module

* #14712: Update composite ops table format

* #14712: Update unary op fast and approximate mode table

* #14712: Update table style for op with float param

* #14712: Update bitwise ops doc

* #14712: Update doc for glu ops

* #14712: Update composite ops

* #14712: Udpate composite ops description

* #14712: Update unary documentation

---------

Co-authored-by: umdevimcw <umdevi@multicorewareinc.com>

sjameelTT committed:
#12548: fix reshape mismatch on Blackhole when input is from DRAM + misaligned (#14766)

umadevimcw committed:
#14713: Remove unnecessary usage of creation ops

npetrovic-tenstorrent committed:
Add new sweeps (#14556)

* #11512: Add round bw sweep

* #11512: Add to run config

umadevimcw committed:
#14713: Remove unnecessary usage of creation ops

amalbasaTT committed:
Amalbasa tt/complex ops sweeps 2 (#14673)

* #11512: Add binary complex add_bw, div_bw

* #11512: Modify invalidate function at binary_complex/div_bw

* #11512: Remove changes from binary_backward/div_bw/div_bw.py

* #11512: Modify ttnn-run-sweeps.yaml file

* #11512: Change complex binary div_bw end shape parameter

umadevimcw committed:
#6991 WHB0 tests file updates

* #6991: Update test skips

* #6991: Rename to update the position of the test files

---------

Co-authored-by: umadevimcw <umadevi@multicorewareinc.com>

KalaivaniMCW committed:
#14734: fix hardtanh golden function (#14736)

umadevimcw committed:
#14714: Remove unnecessary usage of creation ops in binary bw

Aswinmcw committed:
#14406: Add BW computations and t3k perf (#14651)

* #14406: Add BW computations and t3k perf

* #14406: Update BW computation

* #14406: Add shell script for better running

* #14406: Remove py files

* #14406: Compute element_size

* #14406: Fix bfloat dtype bytes

* #14406: Modify csv handling

avoraTT committed:
#14107: Integrate rope op for decode mode in llama (#14755)

yugaoTT committed:
Add block aligned CB sync for matmul use case (#14695)

* #0: add block aligned CB sync for matmul use case

* #0: cleanup

* #0: remove hardcoded tile size

mtairum committed:
Add Llama3.1-70b to the unified llama3 codebase (#14273)

Part 3 of our Llama3 unified codebase

* Single LLama3 codebase now supports 1/3/8/11/70B Llama models.

* Updated CI tests to include the new LLama-70B

* Added all llama models to the models dashboard

* Revamped CI perplexity pipeline and added top1/5 for llama3 models

* Several Llama integration bug fixes and performance tweaks

* Removed old Llama3.1-8B implementation

---------

Co-authored-by: Colman Glagovich <cglagovich@tenstorrent.com>
Co-authored-by: Mark O'Connor <moconnor@tenstorrent.com>
Co-authored-by: Evan Smal <esmalTT@users.noreply.github.com>
Co-authored-by: Johanna Rock <jrock@tenstorrent.com>
Co-authored-by: avoraTT <avora@tenstorrent.com>
Co-authored-by: asaigal <asaigal@tenstorrent.com>

ayerofieiev-tt committed:
Fix sharding handing in TensorLayout and add many tensor tests (#14550)

* #0: Fixes for TensorLayout fromLegacyPaddedShape

* #0: Non-zero page size

* #0: approach fixing sharding handling

* #0: add tests

* #0: fix CMAKE_UNIQUE_NAMESPACE usage

---------

Co-authored-by: Stanislav Minakov <sminakov@tenstorrent.com>

caixunshiren committed:
#0: fixed idle core assignment mismatch for sdpa decode in sharded low batch case

afuller-TT committed:
#0: .clang-tidy checks according to the current state of the repo (#14770)

sankarmanoj-tt committed:
#6326: Implement ConvTranspose2D (#13639)

blozano-tt committed:
#14581: Remove deprecated UMD API host_addr_params (#14774)

broskoTT committed:
UMD bump: Remove usage of create-ethernet-map (#14718)

shwetankTT committed:
#0: testing

eyonland committed:
#0: Fix relative link in export_to_sqlite.py (#14810)

sminakov-tt committed:
#0: Add support for 0-volume and 1-volume tensors for ttnn::add (#14611)

* #0: Add support for 0-volume and 1-volume tensors for ttnn::add

* #0: Fixed an assert

* #0: Tests pass in debug

* #0: Cover more 0-volume and 1-volume cases

* #0: replace TT_ASSERT with TT_FATAL

* #0: Remove 0 size check for circular buffer

blozano-tt committed:
#14623: Remove dead ARCH_NAME specific code

blozano-tt committed:
#0: Remove the include

tt-aho committed:
#0: Add unit test for arbitrary core ordering for sharding

tt-rkim committed:
#14225: Limit log max to 50m to further help with disk space dying for some reason it's still saving to JSON.… (#14805) [skip ci]

* #14225: Limit log max to 50m to further help with disk space dying... for some reason it's still saving to JSON

* #14225: Use local driver

bbeggsTT committed:
Update Tutorial - Add Two Integers in a Baby RISC-V.md (#14814)

tt-dma committed:
#0: Remove DPRINT check on tile idx going out of CB page

Developers prefer to not have to advance the pointer to print tiles, and
instead use tile idx to go past the CB page to print data.

tt-rkim committed:
#0: Add telemetry to fd nightly and single card demos

afuller-TT committed:
#0: Oops. Merged the wrong set (#14824)

THIS is the set that reflects the current state of the repo.  Mea culpa.

blozano-tt committed:
#14580: Remove usage of noc_parameters.h from cq_commands.hpp (#14769)

hschoi4448 committed:
#14707: refactoring return optional tensors (#14708)

* #14707: refactoring return optional tensors

blozano-tt committed:
#14644: Remove dead function from tt_cluster.cpp (#14831)

blozano-tt committed:
#14628: Remove ARCH_NAME specific parameter from kernel.cpp (#14792)

umadevimcw committed:
#0: Fix unbound local error issue (#14795)

umadevimcw committed:
#14781: Update documentation table

umadevimcw committed:
#14783: Update table in binary backward documentation

npetrovic-tenstorrent committed:
Sweeps mean, std, var  (#14793)

* #11512: Add mean, std, var sweeps

* #11512: Fix run config typo

mcw-anasuya committed:
#11086: Update supported params for all ops in unary backward doc (#14376)

* #11086: Update bind_unary_backward_op

* #11086: Update remaining ops

* #14782: Restructure doc for supported params

---------

Co-authored-by: VirdhatchaniKN <virdhatchani.narayanamoorthy@multicorewareinc.com>

mtairum committed:
#0: Disable test_llama_model_prefill from CI

johanna-rock-tt committed:
Add DRAM pre-fetcher micro benchmark (#14850)

* #14512: Add DRAM pre-fetcher benchmark (#14528)

* #14512: Add second core coords for GS and BH

* #14512: Update BW target for GS

* #14512: Remove wrong assert

blozano-tt committed:
#14629: Remove noc_parameters.h inclusion from genfiles.cpp (#14829)

* #14629: Remove noc_parameters.h inclusion from genfiles.cpp

* #0: Fix typo

tt-rkim committed:
#0: Revert "#11795: Update test_pgm_dispatch and sweep" [skip ci]

This reverts commit 6f18219025b9ae27fda7d7ead079e4864a278b29.

This is to fix nightly fast dispatch. Force-merging.

cglagovichTT committed:
#14848: Fix lt exclusion logic. Only apply full text mask in decode (#14853)

#14848: Fix lt exclusion logic. Only apply full text mask if in decode mode.

TT-billteng committed:
New post commit owner (#14858)

#0: new post commit owner

ntarafdar committed:
#13466: Enable Program Cache for Two Concat Tensors 

#13466: fix s2s_rm_concat_two_tensors with program cache

yugaoTT committed:
Add Prefetcher + Matmul ubenchmark (#14811)

* #0: add prefetcher + matmul test

* #0: move common files to the same folder

* #0: add back debug code

* #0: fixes

* #0: move increment_arg_idx to common file

amahmudTT committed:
#14694: Stall unpacker till MMIO write to unpacker address completes

afuller-TT committed:
Afuller/clang tidy ignore not built files (#14856)

#0: Do not check files not present in compile_commands.json

blozano-tt committed:
#14618: Prune unnecessary parameters from set_device_l1_address_params (#14789)

sminakov-tt committed:
#0: Finalize transition of eltwise and matmul operations to SimpleShape (#14777)

* #0: Finalize transition of eltwise and matmul to SimpleShape

* #0: Shape fixup

jbaumanTT committed:
#14867: Fix launch_msg_rd_ptr value sync after trace. (#14869)

When when we enqueue a trace that does not use some set of cores (multicast or
unicast), the launch_msg_rd_ptr isn't reset on those cores. We need to do the
same thing with the worker_launch_message_buffer_state on the host.

blozano-tt committed:
#14582: Remove deprecated UMD API eth_interface_params (#14779)

blozano-tt committed:
[skip ci] #0: Allow Users to provide extra tag for T3K choose your own pipeline (#14873)

mbahnasTT committed:
segformer wh optim and moving to demos folder (#14552)

* #0: Segformer optim and new Create heads OP

blozano-tt committed:
#14633: Fix using incorrect symbol (#14876)

blozano-tt committed:
#14608: l1_address_params init to use Hal (#14872)

Aswinmcw committed:
#14406: Add perf test for reduce  scatter (#14838)

* #14406: Add perf test for reduce  scatter

* #14406: Add perf test for N300 reduce  scatter

* #14406: Fix BW computation

blozano-tt committed:
#0: Quick Fix T3K pipelines (#14881)

Commit:
#14851: Disable shallow conv for tiled input

Disable shallow conv in auto-shard codepath for tiled input.
Too many restrictions from data-movement ops make it hard to support it ATM.
This incluces limits on interleaved2sharded (tiled input, but shard with smaller than 32).
UntializeWithUnpad not being able to work on row-major tensors with odd nubmer of elements.

yieldthought committed:
Yieldthought/lt persist state (#14891)

* #0: Tie name of each log to a slot index

* #0: Detach log id's from visible index as this changes on deletion

* #0: remove logs when deleted from list

* #0: Persist lt status in logs/status.json

* #0: always reset on error code unless cancelled

* #0: Add X to clear all

spoojaryTT committed:
Add device specific defines to kernel build command and hash (#14753)

* #13869: Add defines from device to kernel build hash.

* #13869: Add device specific defines to kernel build command.

* #13869: Move hash calculation of defines to utils.

* #0: Fix issue with changing L1 configuration

ncvetkovicTT committed:
#14900: Disable ComputeCopyBlockSingle Until Solved (#14901)

* #14900: Disable ComputeCopyBlockSingle until solved.

* #14900: Revert seed

nhuang-tt committed:
#14678: add missing defines and args to test_tx_rx (#14679)

* #14678: add missing defines and args to test_tx_rx

* Add missing header to tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tx_rx.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* Add missing header to tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tx_rx.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* Update tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tx_rx.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

wransom-TT committed:
Fix MaxPool Block / Width Sharding with Large Kernels / Wide Reductions (#14531)

* #14249: Fixed bug for width and block sharding with large kernel sizes and wide reductions

afuller-TT committed:
Afuller/clang tidy (#14868)

### Ticket
#14688 

### Problem description
Avoid scanning dependencies or their headers when scanning our code with
clang-tidy.

### What's changed
Placements of no-op .clang-tidy files and suppressing of clang-tidy
itself when building dependencies.
Marking headers of dependencies as `SYSTEM` includes so clang-tidy
doesn't report issues detected in them.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

blozano-tt committed:
[skip ci] Relax commit message requirements (#14904)

### Ticket
NA

### Problem description
Too much friction for contributors.

### What's changed
Relax stringent commit message requirements.

bbradelTT committed:
#13186: adjust negative dims in reduce (#14896)

### Ticket
Link to Github Issue #13186

### Problem description
ttnn.max op throws Unsupported dim error while reducing along the
channel dimension for 4d input and batch dimension for both 4d and 3d
input.

### What's changed
Push the check for the dimension earlier and update the dim array with
the change.

### Checklist
- [x] Post commit CI passes
https://github.com/tenstorrent/tt-metal/actions/runs/11744609127
- [ ] Blackhole Post commit (if applicable) N/A
- [ ] Model regression CI testing passes (if applicable) N/A
- [ ] Device performance regression CI testing passes (if applicable)
N/A
- [x] New/Existing tests provide coverage for changes

sankarmanoj-tt committed:
Allow different shard scheme for input and output tensors in only Width Sharded Conv2d (#14128)

### Ticket
#14123 

### Problem description
Having same shard config for input & output makes the op choose the
largest common core count. Low core count leads to low perf and L1 OOM
errors.

### What's changed
Shard config for Input & Output Tensors can be independent for only
Width Sharded Conv2d.

### Checklist
- [x] Post commit CI
[passes](https://github.com/tenstorrent/tt-metal/actions/runs/11727032596)
- [x] New/Existing tests provide coverage for changes

afuller-TT committed:
Fix CMake re-configure (#14907)

### Ticket
None

### Problem description
I broke re-configure in CMake with an earlier refactor. Lost (or perhaps
no longer inheriting the value from UMD?) for CPM_SOURCE_CACHE. On
re-run CPM would think itself in a different location and early-exit
without providing the functions it was supposed to.

### What's changed
Ensure that our top-level project sets up CPM correctly.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

jvegaTT committed:
14517: making all-gather and barrier support tiny-tiles (#14866)

### Ticket
https://github.com/tenstorrent/tt-metal/issues/14517

### Problem description
Adding the ability for all-gather and barrier to support the following
tiling configurations:
(height, width) =(32,32), (16,32), (8,32), (4,32), (2,32), and (1,32)
For all shapes already supported by all-gather

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

API changes to give the functionalities needed for Tiny Tiles, in
particular with multi-device tensor manipulation
CCL common changes
Added 7 post-commit tests to verify the various configurations in
all-gather. barrier tests put in t3000 frequent as it did not require
any major changes to support tensors with tiny tile
Modifications to the host and kernel code of the all gather

### Checklist
- [x] Post commit CI passes
https://github.com/tenstorrent/tt-metal/actions/runs/11730896322
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable) T3000 frequent
test : https://github.com/tenstorrent/tt-metal/actions/runs/11746021636
passed everywhere except the LLama 2 tests and llm mistral perf tests, but those are failing for
all jobs, seem to be an issue in main or in the testing pipelines.
- [ ] Device performance regression CI testing passes (if applicable)
- [X] New/Existing tests provide coverage for changes

blozano-tt commented on build_metal.sh:
```diff
@@ -155,10 +168,18 @@ echo "INFO: Build tests: $build_tests"
 
 # Prepare cmake arguments
 cmake_args+=("-B" "$build_dir")
-cmake_args+=("-G" "Ninja")
```
Why remove Ninja?

ct-clmsn replied:
This is an artifact of my previous attempt to do a riscv build back in April. Ninja had some issues. I just added that line back into the script and everything is working find. Thanks for finding this one and asking.

ct-clmsn replied:
I've corrected this issue; added ninja back into the script. the PR should be synchronized now.

blozano-tt commented on scripts/build_scripts/build_with_profiler_opt.sh:
```diff
@@ -1,4 +1,9 @@
+# Copyright
+# © 2023 Tenstorrent Inc.
+# © 2024 Tactical Computing Labs, LLC
+#
```
This script was deleted from main. I don't understand why I am seeing it in this diff.

ct-clmsn replied:
This patch was built on a specific commit that is out of sync with main; this commit point was suggested by some folks at TT. This particular script had references to Ninja.

ct-clmsn replied:
I've updated this branch to mirror main; will hopefully have this branch synchronized soon.

ct-clmsn replied:
branch should be synchronized now. I've noticed the cmake scripts require `umd::device` before it's compiled by the build script. I've build umd independent of the build script and added in `find_package` to try to detect an existing build and installation of the dependency.

blozano-tt commented on CMakeLists.txt:
```diff
@@ -191,7 +197,8 @@ ADJUST_COMPILER_WARNINGS()
 add_library(compiler_flags INTERFACE)
 target_link_libraries(compiler_flags INTERFACE linker_flags compiler_warnings stdlib)
-target_compile_options(compiler_flags INTERFACE -mavx2 -fPIC -DFMT_HEADER_ONLY -fvisibility-inlines-hidden -fno-lto)
+#target_compile_options(compiler_flags INTERFACE -mavx2 -fPIC -DFMT_HEADER_ONLY -fvisibility-inlines-hidden -fno-lto)
```
Please cleanup dead code

ct-clmsn replied:
will do, thanks

ct-clmsn replied:
should be removed now

ayerofieiev-tt commented:
Hi @ct-clmsn! This is awesome! Thrilled to see this change coming in! 🤩


We don't have RISCV machines on CI yet. Can you please share what environment do you use to build and test this? 
This might help us to establish some CI.

afuller-TT committed:
#14001: remove header target in CMake (#14748)

### Ticket
#14001 

### Problem description
Globals have a few problems:
* Enables circular dependencies
* Confounds humans and tools from reasoning through the project
structure
* Hinders packaging efforts <-- what prompted this PR

### What's changed
Remove `metal_header_directories` in favour of targets linking against
the actual targets that they depend on, and allowing CMake to flow usage
requirements according to the dependency graph.

### Checklist
- [x] Post commit CI passes
https://github.com/tenstorrent/tt-metal/actions/runs/11689366899
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes
- [x] TTNN C++ Project passes
https://github.com/tenstorrent/tt-metal/actions/runs/11689407001

npetrovic-tenstorrent committed:
Prod backward sweep (#14884)

New sweeps:
- prod_bw

dvartaniansTT committed:
#0: yolov4 bringup tech report

mbezuljTT committed:
GraphProcessor extension to expose page size and num cores (#14903)

This change is part of larger change to support compiler

### Problem description
For L1 allocated buffers, using graph processor, we want to be able to
know what is per core L1 usage.

### What's changed
Addded GraphProcessor functionality for
buffer_allocation/buffer_deallocation nodes to capture page size and
number of cores (for sharded buffer allocations)

Extend graph utilities with
* extract_l1_output_buffer_allocation_size_per_core
* extract_l1_buffer_allocation_peak_size_per_core

Extended graph_test_add to test this functionality. Also heavy tested
locally with different ops and options.


### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

Commit:
synchronizing with main

Commit:
synchronizing with main

Commit:
updated

Commit:
Add transposed tile layout and fix tiny tiles with Bfloat format (#14491)

* #0: fix tiny tile with bfp8 format + add transpose of tiles with float16 format

* #0: add bfloat4 support for tiny tiles

* #0: fix bfloat format tiny tiles

* #0: temp fix, bfloat works, transpose backward pass failed

* #0: fix transpose tile, but pcc is lower

* #0: fix pcc err, matmul 2d 1d failed with transposed tile

* #0: fix tiny tile transposed pcc, fix matmul transposed tile

* #0: add tile conversion for numpy tensor + shorten matmul test duration

* #0: add to_layout for tt to torch tensor path

* #0: fix multi-device bfloat torch to tt tensor

* #0: fix model and sweeps

* #0: fix split qkv

* #0: rm some optional tile arg

* #0: use tenary operator

* #0: add doc string for to_torch

* #0: remove extra tensor creation

* #0: minor fix

* #0: remove comment

* Revert "#0: remove extra tensor creation"

This reverts commit 94796e604cf08d85b6c3f26de765dc144395a9d5.

ct-clmsn commented:
@blozano-tt apologies for the side effects of doing the rebase

joelsmithTT commented on tt_metal/common/bfloat8.hpp:
```diff
@@ -248,12 +284,122 @@ inline std::vector<float> unpack_bfp8_tiles_into_float_vec(const std::vector<uin
                         _mm256_storeu_ps(&float_vec[float_data_index],  _mm256_castsi256_ps(man));
+#else
+			std::vector<std::uint32_t> exp_vector(8, get_byte(exp_word, sub_word_index));
+                        std::vector<std::uint32_t> first(4, bfp8_tiles.at(16 + tile_and_data_index));
+                        std::vector<std::uint32_t> second(4, 16 + tile_and_data_index + 1);
+                        std::vector<std::uint32_t> combined(8, 0);
+
+                        for(std::size_t j = 0; j < 8; ++j) {
+			   if(shift[j] < 32) {
+   			      first[j] = first[j] & mask[j] >> shift[j];
+			   }
+			   else {
+   			      first[j] = 0;
+			   }
+			}
```
It looks like this will perform out-of-bounds access on `first` (size == 4, loop iterates 8 times)

joelsmithTT replied:
tests/tt_metal/tt_metal/test_bfp8_conversion.cpp has a test case for the conversion code in bfloat8.hpp

ct-clmsn replied:
Nice catch, thanks for the code review. I'll make updates this evening.

jvegaTT commented:
I am one of the code owners for the CCL files. We are seeing all the CCL changes from the past few months in the diff in this PR with 92 affected files just in the CCL side, it is impossible to filter out what was actually changed if anything in this PR pertaining to our area. I imagine there is a similar problem being faced by others as there are 2507 changed files! I don't think we can reasonably review this. I suggest reverting to an earlier commit before the erroneous merge with main and to perform a rebase instead.

ct-clmsn commented:
@jvegaTT not a problem; completely understand the situation. Not exactly sure how the PR went sideways but I'll see about getting this fixed this evening. would you all prefer the branch get a hard reset to main or the branch get deleted and replaced with a new branch that's been created from main?

SeanNijjar commented:
> @jvegaTT not a problem; completely understand the situation. Not exactly sure how the PR went sideways but I'll see about getting this fixed this evening. would you all prefer the branch get a hard reset to main or the branch get deleted and replaced with a new branch that's been created from main?

I _think_ GH will need the pr reopened to have all the extra code owners and changes removed from it.

ct-clmsn commented:
Cool, I'll close this and reopen. Thanks.
